src/dc/nipals.js

```diff
@@ -0,0 +1,64 @@
+
+/**
+ * NIPALS
+ * @param {Matrix} X
+ * @param {*} Y
+ * @param {*} u
+ */
+export function nipals(X, options = {}) {
+  const { Y } = options;
+  let u;
+  if (Y) {
+    u = Y.getColumnVector(0);
+  } else {
+    u = X.getColumnVector(0);
+  }
+
+  let diff = 1;
+  let t, q, w, tOld;
+
+  let counter = 0;
+  for (var i = 0; diff > 1e-10; i++) {
+    w = X.transpose().mmul(u).div(u.transpose().mmul(u).get(0, 0));
+    w = w.div(w.norm());
+
+    t = X.mmul(w).div(w.transpose().mmul(w).get(0, 0));
+
+    if (i > 0) {
+      diff = t.clone().sub(tOld).pow(2).sum();
+    }
+    tOld = t.clone();
+
+    if (Y) {
+      q = Y.transpose().mmul(t).div(t.transpose().mmul(t).get(0, 0));
+      q = q.div(q.norm());
+
+      u = Y.mmul(q).div(q.transpose().mmul(q).get(0, 0));
+    } else {
+      u = t;
+    }
+
+    counter++;
+    if (counter > 1000) break;
```

Comment:
I think this should be an option with default value 1000. Not nice to have an hardcoded value in the code.

Reply:
The code has been modified as suggested. It is now an option.maxIterations with default value set to 1000